### ISSUE TITLE
How to call an action from NextJS getInitialProps?

### ISSUE THREAD

**Issue by joshuatvernon** [atlassian/react-sweet-state#86]
Do you have any idea how I can I call an action from within `getInitialProps` in a NextJS app. My action makes a `/graphql` request and then updates the state. Therefore, I need to have that state before the first server side render. 

Here is a component I use in my `_document.tsx` render method to add the state to a script tag. I also have code to parse it and re-hydrate the store on client side. However, I currently get errors when the action is fired outside of the `getInitialProps` since it tries to render again since the state has chanegd.
```
import React, { FunctionComponent } from 'react';

import { useStore } from '../..';
import { isBrowser } from '../../../utils/env';
import { InitialStateScriptProps } from './types';

export const INITIAL_STATE_SCRIPT_NAME = '__INITIAL_STATE__';

/**
 * Get initial state from InitialStateScript
 *
 * @param name (optional) falls back to INITIAL_STATE_SCRIPT_NAME
 */
export const getInitialState = (name?: string) =>
  isBrowser() ? (window as any)[name ? name : INITIAL_STATE_SCRIPT_NAME] : undefined;

export const InitialStateScript: FunctionComponent<InitialStateScriptProps> = props => {
  const { name = INITIAL_STATE_SCRIPT_NAME } = props;
  const [state] = useStore();
  const INITIAL_STATE = JSON.stringify(state);
  return (
    <script
      /* eslint-disable react/no-danger-with-children */
      dangerouslySetInnerHTML={{ __html: `window.${name} = ${INITIAL_STATE};` }}
    />
  );
};
```
I'm not sure whether I should be calling my action in `getInitialProps` or whether I should be doing something like **Apollo GraphQL** where I make all my `/graphql` calls in my `_app.tsx` and then I rewind the Head and then store the result of the call in the store. It's a bit tricky to use Apollo GraphQL, NextJs and ReactSweetState together . . .

**Comment by joshuatvernon** [comment]
I managed to do this by using the `defaultRegistry` e.g.
```
import React from 'react';

import { defaultRegistry } from 'react-sweet-state';
import { NextPage } from 'next';

const MyPage = props => <MyDumbPage />;

MyPage.getInitialProps = async (ctx: NextPageContext) => {
    // Accessing actions with `getInitialProps`
    await defaultRegistry.getStore(store, requestId).actions.myAction();

    // Accessing state within `getInitialProps`
    const state = defaultRegistry.getStore(store, requestId).storeState.getState();

    return {
        myVariable: state.myVariable
    };
}
```

**Comment by albertogasparin** [comment]
Hi @joshuatvernon, sorry for the late reply. Yes, given `getInitialProps` is a static method you have to rely on `defaultRegistry` in order to call the action and populate the store from there. `defaultRegistry` is not fully documented but very unlikely to change :) 

Another option I've used is to make `myAction()` more decomposable, so part of it can be called outside sweet-state, then pass the data into a `Container` so it is accessible to all consumers along below tree:

```jsx
const myAction = () => async ({ setState }) => {
   const data = await fetchData();
   setState({ data });
}

MyPage.getInitialProps = async () => {
  const data = await fetchData();
  return data;
}

function MyPage({ data }) {
  return <MyContainer data={data}>...</MyContainer>
}
```

Basically very similar to what [rehydration docs describe](https://atlassian.github.io/react-sweet-state/#/advanced/rehydration).

**Comment by joshuatvernon** [comment]
Sorry, for the even later response. Using the `registry` worked like a charm. The solution ended up being fairly elegant and may be worth sharing at some stage, however, for now thanks and you can close this issue 🙇‍♂️
